server/src/main/java/org/apache/iotdb/db/mpp/buffer/SinkHandle.java

```diff
@@ -0,0 +1,346 @@
+  private NewDataBlockEvent savedNewDataBlockEvent;
+  private int numOfBufferedTsBlocks = 0;
+  private long bufferRetainedSizeInBytes;
+  private boolean closed;
+  private boolean noMoreTsBlocks;
+  private Throwable throwable;
+
+  public SinkHandle(
+      String remoteHostname,
+      TFragmentInstanceId remoteFragmentInstanceId,
+      String remoteOperatorId,
+      TFragmentInstanceId localFragmentInstanceId,
+      LocalMemoryManager localMemoryManager,
+      ExecutorService executorService,
+      DataBlockService.Client client,
+      TsBlockSerde serde,
+      SinkHandleListener sinkHandleListener) {
+    this.remoteHostname = Validate.notNull(remoteHostname);
+    this.remoteFragmentInstanceId = Validate.notNull(remoteFragmentInstanceId);
+    this.remoteOperatorId = Validate.notNull(remoteOperatorId);
+    this.localFragmentInstanceId = Validate.notNull(localFragmentInstanceId);
+    this.localMemoryManager = Validate.notNull(localMemoryManager);
+    this.executorService = Validate.notNull(executorService);
+    this.client = Validate.notNull(client);
+    this.serde = Validate.notNull(serde);
+    this.sinkHandleListener = Validate.notNull(sinkHandleListener);
+  }
+
+  @Override
+  public ListenableFuture<Void> isFull() {
+    if (closed) {
+      throw new IllegalStateException("Sink handle is closed.");
+    }
+    return nonCancellationPropagating(blocked);
+  }
+
+  private void submitSendNewDataBlockEventTask(int startSequenceId, List<Long> blockSizes) {
+    executorService.submit(new SendNewDataBlockEventTask(startSequenceId, blockSizes));
+  }
+
+  @Override
+  public void send(List<TsBlock> tsBlocks) throws IOException {
+    Validate.notNull(tsBlocks, "tsBlocks is null");
+    if (throwable != null) {
+      throw new IOException(throwable);
+    }
+    if (closed) {
+      throw new IllegalStateException("Sink handle is closed.");
+    }
+    if (!blocked.isDone()) {
+      throw new IllegalStateException("Sink handle is blocked.");
+    }
+    if (noMoreTsBlocks) {
+      return;
+    }
+
+    long retainedSizeInBytes = 0L;
+    for (TsBlock tsBlock : tsBlocks) {
+      retainedSizeInBytes += tsBlock.getRetainedSizeInBytes();
+    }
+    int currentEndSequenceId;
+    List<Long> tsBlockSizes = new ArrayList<>();
+    synchronized (this) {
+      currentEndSequenceId = bufferedTsBlocks.size();
+      blocked =
+          localMemoryManager
+              .getQueryPool()
+              .reserve(localFragmentInstanceId.getQueryId(), retainedSizeInBytes);
+      bufferRetainedSizeInBytes += retainedSizeInBytes;
+      bufferedTsBlocks.addAll(tsBlocks);
+      numOfBufferedTsBlocks += tsBlocks.size();
+      for (int i = currentEndSequenceId; i < currentEndSequenceId + tsBlocks.size(); i++) {
+        tsBlockSizes.add(bufferedTsBlocks.get(i).getRetainedSizeInBytes());
+      }
+    }
+
+    submitSendNewDataBlockEventTask(currentEndSequenceId, tsBlockSizes);
```

Comment:
Currently, each time we will only send one TsBlock, if in this way, each calling `send` will submit one `SendNewDataBlockEvent` causing a network communication which is not very efficient.

Reply:
It is a dilemma of network efficiency or query latency. The current implementation trades the network for better latency.

Reply:
Also, the size of a TsBlock should be limited by an upper bound, like 1MB. But I believe the sink operator may produce more than that at each time. What do you think?

Reply:
TODO added.